### DISCUSSION TITLE
How to have first worker initialize differently from 2/3/4

### DISCUSSION THREAD

**Discussion by inspire22** [fastapi/fastapi#12793] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List, Any
import numpy as np
from fastapi import FastAPI, Response
from fastapi.responses import ORJSONResponse
from pydantic import BaseModel
import orjson
app = FastAPI()

# uses my encoder.py still from before...
from quaterion_models import SimilarityModel

FOR ONLY ONE OF THE WORKERS:
model = SimilarityModel.load("./my1").to('cuda')

FOR ALL THE OTHER WORKERS:
model = SimilarityModel.load("./my1").to('cpu')
```

### Description

I'm serving LLM embeddings, but have only one GPU. I want to customize the first uvicorn instance to call myModel.to('cuda') and the others myModel.to('cpu').

I looked for environment variables uvicorn (or whatever fastapi run --workers 4 q_api.py uses) might pass when each worker is initialized, but I couldn't find anything like that.

I looked at lifecycle callbacks, but it doesn't seem to be different for each worker.

### Operating System

Linux

### FastAPI Version

0.115.4

### Pydantic Version

2.7.1

### Python Version

Python 3.8.10 ( required by quaterion )

### Additional Context

Thanks for any suggestions!

**Comment by YuriiMotov** [comment, marked as answer]
One way is to use redis lock.
So, the worker which acquired the lock should call `myModel.to('cuda')` and others should call `myModel.to('cpu')`.

```
from contextlib import asynccontextmanager
import uuid

from fastapi import FastAPI
import redis.asyncio as redis

REDIS_HOST = "localhost"
REDIS_PORT = 6379
REDIS_DB = 1

LOCK_KEY = "MODEL_TO_CUDA"


@asynccontextmanager
async def lifespan(app: FastAPI):
    worker_uuid = uuid.uuid4().hex
    redis_lock_client = redis.Redis(host=REDIS_HOST, port=REDIS_PORT, db=REDIS_DB)
    lock_acquired = await redis_lock_client.set(LOCK_KEY, worker_uuid, nx=True)
    if lock_acquired:
        print("CUDA")
    else:
        print("CPU")

    try:
        yield
    finally:
        if lock_acquired:
            await redis_lock_client.delete(LOCK_KEY)


app = FastAPI(lifespan=lifespan)

```

There might be a situation when your process fails without executing `finally` block. In this case the lock will remain set and workers will not be able to acquire it. If it happens, you can release the lock manually. Or you can change the logic of code to handle this situation

**Reply by inspire22** [reply]
Wouldn't this code just re-set the key for each worker?  I'd expect it to use redis.lock, though am new enough to python reading this makes my eyes bleed.  https://redis-py.readthedocs.io/en/stable/lock.html

**Reply by YuriiMotov** [reply]
No, the code will only set lock value if it's not set yet. `nx=True` parameter means "if not exists".

But there were a mistake in my code with removing the lock. I edited it, added a condition so that only the worker which acquired the lock will release it.

**Comment by inspire22** [comment]
What an interesting use of that, thanks :)

--
Thanks,
Kevin
'One way to hold back the passage of time is to make each day the same.' -
Philip Larkin

**Comment by jgould22** [comment]
Are all "workers"  on the same machine (it sounds like it I am just clarifying) ?

If all workers are on the same machine you can use a lock file. Have workers check for a specific file and if there is not one active it can claim the gpu, and if there is one it can default to cpu.

There are some quirks to lock files though so you may want to use a library that takes those quirks into account.

Something like this https://py-filelock.readthedocs.io/en/latest/index.html (though I have never used this lib and cannot vouch for it)

**Reply by inspire22** [reply]
Yes, thanks!  Looks similar to the other idea though - and I'm more familiar with redis than file locks :)  Thanks for the help

**Comment by inspire22** [comment]
Thanks!
--
Thanks,
Kevin
'One way to hold back the passage of time is to make each day the same.' -
Philip Larkin
